Crash when exiting

Note: this might reference #715.

I've built Notes from source a few minutes ago to give it a run and stumbled upon both a bug and a crash that might be related to it. When I open the settings they are not rendered properly as can be seen below. Furthermore, the window cannot be moved since it closes immediately when clicking anywhere but in content area. Clicking in the area where my mouse is apparently counts as 'in content area' since the window does not close.
![image](https://github.com/user-attachments/assets/f2a2a5ec-3c9b-4aed-87e9-86909cae11c0)

So much on the bug.

The crash occours every time when closing the application after the settings window has been used (this might relate to #715). 
I have no experience with gdb, all I can provide for now is
```
Thread 1 "notes" received signal SIGSEGV, Segmentation fault.
0x00007ffff5746804 in QWindow::parent(QWindow::AncestorMode) const ()
    from /lib64/libQt6Gui.so.6
```

Build output (stripped):
```
$ cmake .. -DCMAKE_BUILD_TYPE=Debug
-- The CXX compiler identification is GNU 14.2.1
-- Performing Test HAVE_STDATOMIC - Success
-- Found WrapAtomic: TRUE
-- Found OpenGL: /usr/lib64/libOpenGL.so
-- Found WrapOpenGL: TRUE
-- Found XKB: /usr/lib64/libxkbcommon.so (found suitable version "1.7.0", minimum required is "0.5.0")
-- Found WrapVulkanHeaders: /usr/include
-- Success! Configuration details:
-- App name: Notes
-- App version: 2.3.1
-- Qt version: 6.8.1
-- Update checker: ON
-- Pro Version: ON
-- Build type: Debug
CMake Warning (dev) at /usr/lib64/cmake/Qt6Qml/Qt6QmlMacros.cmake:3051 (message):
  src/qml/Utilities.js is not an ECMAScript module and also doesn't contain
  '.pragma library'.  It will be re-evaluated in the context of every QML
  document that explicitly or implicitly imports io.github.nuttyartist.notes.
  Set its QT_QML_SKIP_QMLDIR_ENTRY source file property to FALSE if you
  really want this to happen.  Set it to TRUE to prevent it.
Call Stack (most recent call first):
  /usr/lib64/cmake/Qt6Qml/Qt6QmlMacros.cmake:896 (qt6_target_qml_sources)
  /usr/lib64/cmake/Qt6Qml/Qt6QmlMacros.cmake:1232 (qt6_add_qml_module)
  CMakeLists.txt:328 (qt_add_qml_module)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found X11: /usr/include
-- Looking for XOpenDisplay in /usr/lib64/libX11.so;/usr/lib64/libXext.so
-- Looking for XOpenDisplay in /usr/lib64/libX11.so;/usr/lib64/libXext.so - found
-- Looking for gethostbyname - found
-- Looking for connect - found
-- Looking for remove - found
-- Looking for shmat - found
```

Make yields the following warning (stripped absolute path to notes folder):
```
notes/src/mainwindow.cpp: In member function ‘virtual bool MainWindow::eventFilter(QObject*, QEvent*)’:
notes/src/mainwindow.cpp:3921:42: warning: ‘static void QApplication::setActiveWindow(QWidget*)’ is deprecated: Use QWidget::activateWindow() instead. [-Wdeprecated-declarations]
 3921 |             QApplication::setActiveWindow(
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
 3922 |                     this); // TODO: The docs say this function is deprecated but it's the only one
      |                     ~~~~~                 
In file included from /usr/include/qt6/QtWidgets/QApplication:1,
                 from notes/build/Notes_autogen/include/ui_mainwindow.h:13,
                 from notes/src/mainwindow.cpp:8:
/usr/include/qt6/QtWidgets/qapplication.h:84:17: note: declared here
   84 |     static void setActiveWindow(QWidget* act);
      |    
```

I run Fedora 41 with GNOME DE (which should provide wayland? cmake found x11 however).

--- a/.clang-format
+++ b/.clang-format
@@ -106,3 +106,7 @@ AlignEscapedNewlines: Left
 # initializer list in list-initialization.
 SpaceBeforeCpp11BracedList: false
 
+# Custom styles added by this project
+
+# fix for https://bugreports.qt.io/browse/QTCREATORBUG-31858
+StatementAttributeLikeMacros: [emit, Q_EMIT]

--- a/.github/workflows/format-lint.yml
+++ b/.github/workflows/format-lint.yml
@@ -7,7 +7,9 @@ on:
 jobs:
   lint:
     name: Format / Lint
-    runs-on: ubuntu-latest
+    # FIXME: ubuntu-latest has clang-format v18 which contains a bug: # https://bugreports.qt.io/browse/QTCREATORBUG-31858. 
+    # Update this once ubuntu ships clang-format v19
+    runs-on: ubuntu-22.04 
     strategy:
       fail-fast: false
     steps:

--- a/.github/workflows/linux.yml
+++ b/.github/workflows/linux.yml
@@ -266,7 +266,7 @@ jobs:
 
   snap:
     name: snap
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-22.04 # FIXME: See this issue before updating to ubuntu-24.08: https://github.com/actions/runner-images/issues/9932
     steps:
       - name: Checkout code
         uses: actions/checkout@v4

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -17,6 +17,7 @@ option(GIT_REVISION
 option(UPDATE_CHECKER
        "Enable or disable both the update checker and auto-updater" ON)
 option(PRO_VERSION "Enable or disable Notes Pro features" ON)
+option(ENABLE_ASAN "Enable address sanitizer" OFF)
 
 project(
   Notes
@@ -31,6 +32,13 @@ set(CMAKE_INCLUDE_CURRENT_DIR ON)
 set(CMAKE_AUTOMOC ON)
 set(CMAKE_AUTOUIC ON)
 set(CMAKE_AUTORCC ON)
+
+# address sanitizer can be enabled by running cmake with -DENABLE_ASAN=ON
+if(ENABLE_ASAN AND (CMAKE_CXX_COMPILER_ID STREQUAL "GNU"
+                    OR CMAKE_CXX_COMPILER_ID MATCHES "Clang"))
+  set(ADDRESS_SANITIZER_FLAGS "-fsanitize=address -fno-omit-frame-pointer")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${ADDRESS_SANITIZER_FLAGS}")
+endif()
 
 if(POLICY CMP0083)
   cmake_policy(SET CMP0083 NEW)
@@ -72,6 +80,10 @@ if(CMAKE_BUILD_TYPE STREQUAL "")
   message(STATUS "Build type: (not set)")
 else()
   message(STATUS "Build type: ${CMAKE_BUILD_TYPE}")
+endif()
+
+if(ENABLE_ASAN)
+  message(STATUS "Enabling AddressSanitizer")
 endif()
 
 set(GIT_REV "")

--- a/docs/build_options.md
+++ b/docs/build_options.md
@@ -10,6 +10,7 @@ You should set those while invoking CMake to build the project. See the [example
 | `GIT_REVISION`                             | `OFF`         | `ON` / `OFF`        | Append the current git revision to the app's version string |
 | `UPDATE_CHECKER`                           | `ON`          | `ON` / `OFF`        | Enable or disable both the update checker and auto-updater  |
 | `PRO_VERSION`                              | `ON`          | `ON` / `OFF`        | Enable or disable Notes Pro features                        |
+| `ENABLE_ASAN`                              | `OFF`         | `ON` / `OFF`        | Enable AddressSanitizer (ASan) for debugging                |
 
 ### Examples
 

--- a/src/dbmanager.cpp
+++ b/src/dbmanager.cpp
@@ -2502,9 +2502,8 @@ void DBManager::exportNotes(const QString &baseExportPath, const QString &extens
         counter = 1;
         while (directory.exists(filePath)) {
             filePath = QStringLiteral("%1%2%3 %4%5")
-                               .arg(notePath, QDir::separator(), safeTitle)
-                               .arg(counter++)
-                               .arg(extension);
+                               .arg(notePath, QDir::separator(), safeTitle,
+                                    QString::number(counter++), extension);
         }
 
         // qDebug() << "Exporting note:" << filePath;

--- a/src/mainwindow.cpp
+++ b/src/mainwindow.cpp
@@ -1047,7 +1047,7 @@ void MainWindow::setupEditorSettings()
     m_editorSettingsQuickView.setResizeMode(QQuickView::SizeViewToRootObject);
     m_editorSettingsQuickView.setFlags(Qt::FramelessWindowHint);
     m_editorSettingsQuickView.setColor(Qt::transparent);
-    m_editorSettingsWidget = QWidget::createWindowContainer(&m_editorSettingsQuickView, nullptr);
+    m_editorSettingsWidget = QWidget::createWindowContainer(&m_editorSettingsQuickView, this);
 #if defined(Q_OS_MACOS)
 #  if QT_VERSION >= QT_VERSION_CHECK(6, 2, 0)
     m_editorSettingsWidget->setWindowFlags(Qt::Popup | Qt::FramelessWindowHint
@@ -1291,7 +1291,7 @@ void MainWindow::setupKanbanView()
     m_kanbanQuickView.rootContext()->setContextProperty("mainWindow", this);
     m_kanbanQuickView.setSource(source);
     m_kanbanQuickView.setResizeMode(QQuickView::SizeRootObjectToView);
-    m_kanbanWidget = QWidget::createWindowContainer(&m_kanbanQuickView);
+    m_kanbanWidget = QWidget::createWindowContainer(&m_kanbanQuickView, this);
     m_kanbanWidget->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
     m_kanbanWidget->hide();
     ui->verticalLayout_textEdit->insertWidget(ui->verticalLayout_textEdit->indexOf(m_textEdit),

--- a/src/mainwindow.cpp
+++ b/src/mainwindow.cpp
@@ -3918,10 +3918,7 @@ bool MainWindow::eventFilter(QObject *object, QEvent *event)
     }
     case QEvent::ActivationChange: {
         if (m_editorSettingsWidget->isHidden()) {
-            QApplication::setActiveWindow(
-                    this); // TODO: The docs say this function is deprecated but it's the only one
-                           // that works in returning the user input from m_editorSettingsWidget
-                           // Qt::Popup
+            QWidget::activateWindow();
             m_textEdit->setFocus();
         }
         break;

--- a/src/mainwindow.cpp
+++ b/src/mainwindow.cpp
@@ -129,7 +129,7 @@ MainWindow::MainWindow(QWidget *parent)
                                         "master/notes_purchase_data.json")),
       m_purchaseDataAlt2(
               QStringLiteral("https://www.rubymamistvalove.com/notes/notes_purchase_data.json")),
-      m_dataBuffer(new QByteArray()),
+      m_dataBuffer(std::make_unique<QByteArray>()),
       m_netManager(new QNetworkAccessManager(this)),
       m_reqAlt1(QNetworkRequest(QUrl(m_purchaseDataAlt1))),
       m_reqAlt2(QNetworkRequest(QUrl(m_purchaseDataAlt2))),
@@ -1737,12 +1737,12 @@ void MainWindow::setupDatabases()
 void MainWindow::setupModelView()
 {
     m_listView = ui->listView;
-    m_tagPool = new TagPool(m_dbManager);
+    m_tagPool = new TagPool(m_dbManager, this);
     m_listModel = new NoteListModel(m_listView);
     m_listView->setTagPool(m_tagPool);
     m_listView->setModel(m_listModel);
     m_listViewLogic = new ListViewLogic(m_listView, m_listModel, m_searchEdit, m_clearButton,
-                                        m_tagPool, m_dbManager, this);
+                                        m_tagPool, m_dbManager, m_listView);
     m_treeView = ui->treeView;
     m_treeView->setModel(m_treeModel);
     m_treeViewLogic = new TreeViewLogic(m_treeView, m_treeModel, m_dbManager, m_listView, this);

--- a/src/mainwindow.h
+++ b/src/mainwindow.h
@@ -262,7 +262,7 @@ private:
     QWindow *m_subscriptionWindow;
     QString m_purchaseDataAlt1;
     QString m_purchaseDataAlt2;
-    QByteArray *m_dataBuffer;
+    std::unique_ptr<QByteArray> m_dataBuffer;
     QNetworkAccessManager *m_netManager;
     QNetworkRequest m_reqAlt1;
     QNetworkRequest m_reqAlt2;

--- a/src/qml/EditorSettings.qml
+++ b/src/qml/EditorSettings.qml
@@ -47,7 +47,7 @@ Item {
             var settingsPaneHeightByParentWindow = 0.80 * data.parentWindowHeight; // 80 percent of the parent window's height
             settingsPane.height = scrollViewControl.contentHeight > settingsPaneHeightByParentWindow ? settingsPaneHeightByParentWindow : scrollViewControl.contentHeight;
             revealSettingsAnimation.start();
-            settingsContainer.upadteScrollBarPosition();
+            settingsContainer.updateScrollBarPosition();
         }
 
         function onMainWindowResized (data) {
@@ -164,7 +164,7 @@ Item {
         easing.type: Easing.InOutQuad
     }
 
-    function upadteScrollBarPosition () {
+    function updateScrollBarPosition () {
         editorSettingsVerticalScrollBar.position = settingsContainer.latestScrollBarPosition;
     }
 

--- a/src/qml/Utilities.js
+++ b/src/qml/Utilities.js
@@ -1,3 +1,4 @@
+.pragma library
 // It takes four arguments. The verb, which defines the HTTP verb to be used (GET, POST, PUT, DELETE).
 // The second paramater is the BASE address (e.g. ‘http://localhost:5000 (opens new window)’).
 // The third parameter is the endpoint to be used as a postfix to the BASE address

--- a/src/qml/kanbanMain.qml
+++ b/src/qml/kanbanMain.qml
@@ -789,7 +789,7 @@ Item {
 
                             onVisibleChanged: {
                                 if (editorSettingsPopupContainer.visible) {
-                                    editorSettings.upadteScrollBarPosition();
+                                    editorSettings.updateScrollBarPosition();
                                 }
                             }
                         }

--- a/src/treeviewlogic.cpp
+++ b/src/treeviewlogic.cpp
@@ -384,8 +384,22 @@ void TreeViewLogic::onMoveNodeRequested(int nodeId, int targetId)
     NodeData target;
     QMetaObject::invokeMethod(m_dbManager, "getNode", Qt::BlockingQueuedConnection,
                               Q_RETURN_ARG(NodeData, target), Q_ARG(int, targetId));
+    // only allow moving a node into a folder
     if (target.nodeType() != NodeData::Folder) {
         qDebug() << __FUNCTION__ << "Target is not folder!";
+        return;
+    }
+    // don't allow moving a node into itself (not sure how this can ever happen but just in case)
+    if (nodeId == targetId) {
+        qDebug() << __FUNCTION__ << "Can't move a node into itself";
+        return;
+    }
+    // don't allow moving a node into the same parent
+    NodeData node;
+    QMetaObject::invokeMethod(m_dbManager, "getNode", Qt::BlockingQueuedConnection,
+                              Q_RETURN_ARG(NodeData, node), Q_ARG(int, nodeId));
+    if (node.parentId() == targetId) {
+        qDebug() << __FUNCTION__ << "Can't move a node into the same parent";
         return;
     }
     emit requestMoveNodeInDB(nodeId, target);

